### ISSUE TITLE
Removing the container at the end of Run-AlPipeline

### ISSUE THREAD

**Issue by RadoArvay** [microsoft/navcontainerhelper#2826]
Please I need an advice. Maybe I am missing something (a parameter or a step) only:

I have an Azure DevOps agent to run pipelines where `Run-AlPipeline` is used. Sometimes I get the error "_The process cannot access the file 'C:\AzureAgent\_work\nn\Apps' because it is being used by another process._" at the beginning of a pipeline run (there is the update of files from Git repository and other initial steps). I have found out the previous pipeline inside of `Run-AlPipeline` has ended up with an error and the container has remained running (there is a condition `if (!($err))` before "Removing container" script part).
After that, I need to stop (and remove - probabelly not needed) the container manually. I am not sure it happens every time the error occures.

Do I need to add an extra step to the pipelines to stop the container forcely after `Run-AlPipeline` or can I use any parameter to prevent the container remain/running?

**Comment by jonaswre** [comment]
```
  - task: PowerShell@2
    displayName: 'Cleanup'
    condition: always()
    inputs:
      targetType: filePath
      filePath: 'scripts\Cleanup.ps1'
      arguments: '-environment "AzureDevOps"'
      failOnStderr: false
```

This is a snipped that I copied from the old helloworld


And this is the Cleanup.ps1 maybe this helps 
```
Param(
    [Parameter(Mandatory=$false)]
    [ValidateSet('Local', 'AzureDevOps', 'GithubActions', 'GitLab')]
    [string] $environment = 'Local'
)

. (Join-Path $PSScriptRoot "Read-Settings.ps1") -environment $environment

if ("$AgentName" -ne "Hosted Agent" -and "$AgentName" -notlike "Azure Pipelines*") {
    . (Join-Path $PSScriptRoot "Install-BcContainerHelper.ps1") -bcContainerHelperVersion $bcContainerHelperVersion

    $cleanupMutexName = "Cleanup"
    $cleanupMutex = New-Object System.Threading.Mutex($false, $cleanupMutexName)
    try {
        try {
            if (!$cleanupMutex.WaitOne(1000)) {
                Write-Host "Waiting for other process to finish cleanup"
                $cleanupMutex.WaitOne() | Out-Null
                Write-Host "Other process completed"
            }
        }
        catch [System.Threading.AbandonedMutexException] {
           Write-Host "Other process terminated abnormally"
        }
    
        Remove-BcContainer -containerName $containerName
        Flush-ContainerHelperCache -KeepDays 2
        
        Remove-Module BcContainerHelper
        $path = Join-Path $ENV:Temp $containerName
        if (Test-Path $path) {
            Remove-Item $path -Recurse -Force
        }
    }
    finally {
        $cleanupMutex.ReleaseMutex()
    }
}

```

**Comment by RadoArvay** [comment]
Yes, I have inspired myself by _HelloWorld_ project.
In the pipeline, I have this:
```yaml
  - job: BuildApps
    workspace:
      clean: all
```
and I use `Cleanup.ps1` script with small changes. But `Cleanup` has nothing with the container, I believe.

**Comment by RadoArvay** [comment]
Anyway, the question is still there: why the container is not removed in `Run-AlPipeline` at the end of run when an error occures? Is it a good concept?

**Comment by jonaswre** [comment]
You might want to inspect the container if the pipeline failed. Additionally the method is not only used for your typical CI Pipeline but also to create local development environments. 

Plus I think it's more resistant to clean up in a different task. Because no matter what happens in the Pipeline the Cleanup script always runs at the end. 
I've never encountered any issues with this.

**Comment by freddydk** [comment]
Thanks @jonaswre for replying - it is correct, that Run-AlPipeline is just a command, which can do most of the build process - but many things can happen, which makes the most secure way for cleanup to be a separate task.

BTW - It is not recommended for people to spend time on building pipelines and setup DevOps themselves.
Microsoft recommends AL-Go for GitHub, which is plug and play DevOps for free - and probably have all the functionality you need out of the box.
go to https://aka.ms/al-go and read the readme.

If you do not want GitHub, but need to be on Azure DevOps, then we recommend ALOps or Alpaca, which both are paid solutions.

**Comment by RadoArvay** [comment]
Yeah, I am aware of possible solutions. We use Azure DevOps not only for pipelines.
At first, I created a pipeline with older BcContainerHelper commands and we used it for some purposes a few months, maybe a year or more. After that, I decided to simplify the scripts by using Run-ALxyz commands. It works and we use them a few months. Maybe in the next stage when we will need more complex solution, we look at other one, but BcContainerHelper is a great help for everyone ;-).
Thanks both of you. I am going to update may cleaning scripts.
